mitodl/mitxpro
# Adds tax rate calculation support

#### Pre-Flight checklist


#### What are the relevant tickets?

https://github.com/mitodl/hq/issues/2374\

#### What's this PR do?

Adds tax rate calculation to the ecommerce system. This is in several parts:

- Adds an app to translate IP addresses to assigned country names. This depends on the MaxMind GeoIP databases (it will work with the free GeoLite2 dataset, which is the recommended one). 
- Adds a storage table for tax rate data in the app.
- Adds some hooks to the relevant portions of the ecommerce basket and order APIs to calculate tax if the learner meets the requirements for tax.
- Adds some template tags - `calculate_tax` and `calculate_tax_total` in `mail` - to assist with adding the tax information to the relevant email templates.

This does _not_ include the frontend or email template changes but the data should be there for this - namely, tax rate information has been added to the basket and receipt APIs so that it can be calculated. (The frontend calculates other things, like total cost, so for consistency I just added the rate itself in rather than calculating it in the API call.) 

#### How should this be manually tested?

You will need a copy of the MaxMind GeoLite2 database for best results. This data can be retrieved here:  https://dev.maxmind.com/geoip/geolite2-free-geolocation-data 
You will need a free MaxMind account for this. (If you happen to have access to the more complete GeoIP2 dataset, that'll work too but that one requires payment.) 
 
You will need to add in a tax rate, and a user that should belong to that country. For testing, I added "IN" (India) with 18% GST - this can be done through the Django Admin and will result in a record that matches what we're actually going to have to collect. 

Finally, in your `.env`, set `ECOMMERCE_FORCE_PROFILE_COUNTRY=True` . Otherwise, you won't qualify for taxation - your IP won't be one assigned to an Indian carrier. (Setting this to True just looks at the account profile.) 

To test, add a product to your cart and check out. The CyberSource screens should reflect the price with tax, and looking at the payloads in the Network tab of the inspector should show tax being calculated. The Order object should store information about the tax assessed as well (rate, country, name). The receipt API should additionally include tax information when the order has been completed. 

#### Any background context you want to provide?

Tax is assessed if the following conditions are _both true_:
- The learner has specified that they live in a country that we have a TaxRate object for
- The learner is connecting to the site using an IP address that has been assigned to the same country the learner lives in (as indicated in their profile)

Notably, these conditions would cause the tax calculation to not assess tax:
- Users living in a taxable country but accessing the site from a different physical location (for example, on vacation in another country)
- Users using a VPN or anonymizer service (e.g. Tor) to connect to the site, and therefore doing so with an IP other than the one they've been assigned by their local ISP
